src/condition/lfi_detector.cpp

```diff
@@ -19,93 +17,40 @@ namespace {
-                    child.parameterName, static_cast<std::size_t>(child.parameterNameLength)};
-                if (path.find(key) != std::string_view::npos) {
-                    strings.emplace_back(key);
-                }
-            }
-
-            if (object::is_container(&child)) {
-                if (stack.size() < object_limits::default_max_container_depth) {
-                    stack.push(&child);
-                }
-                continue;
-            }
-
-            if (child.type == DDWAF_OBJ_STRING && child.nbEntries >= min_str_len) {
-                const std::string_view value{
-                    child.stringValue, static_cast<std::size_t>(child.nbEntries)};
-                if (path.find(value) != std::string_view::npos) {
-                    strings.emplace_back(value);
-                }
-            }
-        }
-    }
-
-    return strings;
-}
+using lfi_result = std::optional<std::pair<std::string, std::vector<std::string>>>;
 
-std::vector<std::string_view> split(std::string_view str, char sep)
+lfi_result lfi_impl(std::string_view path, const ddwaf_object &params,
+    const exclusion::object_set_ref &objects_excluded, const object_limits &limits,
+    ddwaf::timer &deadline)
 {
-    std::vector<std::string_view> components;
-
-    std::size_t start = 0;
-    while (start < str.size()) {
-        const std::size_t end = str.find(sep, start);
-
-        if (end == start) {
-            // Ignore zero-sized strings
-            start = end + 1;
+    object::kv_iterator it(&params, {}, objects_excluded, limits);
+    for (; it; ++it) {
+        if (deadline.expired()) {
+            throw ddwaf::timeout_exception();
         }
 
-        if (end == std::string_view::npos) {
-            // Last element
-            components.emplace_back(str.substr(start));
-            start = str.size();
-        } else {
-            components.emplace_back(str.substr(start, end - start));
-            start = end + 1;
+        const ddwaf_object &param = *(*it);
+        if (param.type != DDWAF_OBJ_STRING || param.stringValue == nullptr ||
+            param.nbEntries < min_str_len) {
+            continue;
         }
-    }
 
-    return components;
-}
-
-std::pair<bool, std::string> lfi_impl(std::string_view path, const ddwaf_object &params)
-{
-    auto matched_params = match_params(path, params);
-    if (matched_params.empty()) {
-        return {};
-    }
+        std::string_view value{param.stringValue, static_cast<std::size_t>(param.nbEntries)};
+        if (path.find(value) == std::string_view::npos) {
+            continue;
+        }
 
-    for (const auto &param : matched_params) {
-        if (param[0] == '/' && param == path) {
-            return {true, std::string(param)};
+        if (value[0] == '/' && value == path) {
+            return {{std::string(value), it.get_current_path()}};
         }
 
-        if (!path.ends_with(param)) {
+        if (!path.ends_with(value)) {
```

Comment:
This one is a bit scary, but trying to remove this (for instance, if `value` is a relative path and contains `../`) would almost certainly yield false positives. Having a way to optionally tweak this condition would be valuable.

Reply:
As discussed this isn't an issue.